lib/stdlib/src/supervisor.erl

```diff
@@ -920,65 +918,59 @@ do_terminate(_Child, _SupName) ->
 %% supervisor. 
 %% Returns: ok | {error, OtherReason}  (this should be reported)
 %%-----------------------------------------------------------------
-shutdown(Pid, brutal_kill) ->
-    case monitor_child(Pid) of
-	ok ->
-	    exit(Pid, kill),
-	    receive
-		{'DOWN', _MRef, process, Pid, killed} ->
-		    ok;
-		{'DOWN', _MRef, process, Pid, OtherReason} ->
-		    {error, OtherReason}
-	    end;
-	{error, Reason} ->      
-	    {error, Reason}
+shutdown(#child{pid=Pid, shutdown=brutal_kill} = Child) ->
+    Mon = monitor(process, Pid),
+    exit(Pid, kill),
+    receive
+        {'DOWN', Mon, process, Pid, Reason0} ->
+            case unlink_flush(Pid, Reason0) of
+                killed ->
+                    ok;
+                normal when not (?is_permanent(Child)) ->
+                    ok;
+                Reason ->
+                    {error, Reason}
+            end
     end;
-shutdown(Pid, Time) ->
-    case monitor_child(Pid) of
-	ok ->
-	    exit(Pid, shutdown), %% Try to shutdown gracefully
-	    receive 
-		{'DOWN', _MRef, process, Pid, shutdown} ->
-		    ok;
-		{'DOWN', _MRef, process, Pid, OtherReason} ->
-		    {error, OtherReason}
-	    after Time ->
-		    exit(Pid, kill),  %% Force termination.
-		    receive
-			{'DOWN', _MRef, process, Pid, OtherReason} ->
-			    {error, OtherReason}
-		    end
-	    end;
-	{error, Reason} ->      
-	    {error, Reason}
+shutdown(#child{pid=Pid, shutdown=Time} = Child) ->
+    Mon = monitor(process, Pid),
+    exit(Pid, shutdown),
+    receive
+        {'DOWN', Mon, process, Pid, Reason0} ->
+            case unlink_flush(Pid, Reason0) of
+                shutdown ->
+                    ok;
+                {shutdown, _} ->
```

Comment:
As {shutdown, Reason} is a way for the child process to terminate itself normally (gracefully but with a trail as to why). Maybe it should also have a guard not (is_pernmanent(Child)). I must confess I actually have not used permanent children. But that would be backwards compatible and it seems to make sense that if we do not accept normal as a shutdown reason {shutdown, Reason} should not be accepted either and only the shutdown initiated by the supervisor should be considered a non-fault.

Reply:
Makes sense I think.

Reply:
Changed as suggested in last commit.

Reply:
👍